src/settings/PathConfigStorage.cpp

```diff
@@ -27,209 +28,209 @@ std::vector<double> PathConfigStorage::getLineWidthFactorPerExtruder(const Slice
-    , mesh.getSettingInPercentage("bridge_skin_material_flow_3")
-    , GCodePathConfig::SpeedDerivatives{mesh.getSettingInMillimetersPerSecond("bridge_skin_speed_3"), mesh.getSettingInMillimetersPerSecond("acceleration_topbottom"), mesh.getSettingInMillimetersPerSecond("jerk_topbottom")}
+    , mesh.settings.get<Ratio>("bridge_skin_material_flow_3")
+    , GCodePathConfig::SpeedDerivatives{mesh.settings.get<Velocity>("bridge_skin_speed_3"), mesh.settings.get<Velocity>("acceleration_topbottom"), mesh.settings.get<Velocity>("jerk_topbottom")}
     , true // is_bridge_path
-    , mesh.getSettingInPercentage("bridge_fan_speed_3")
+    , mesh.settings.get<Ratio>("bridge_fan_speed_3") * 100.0
 )
 , roofing_config(
     PrintFeatureType::Skin
-    , mesh.getSettingInMicrons("roofing_line_width")
+    , mesh.settings.get<coord_t>("roofing_line_width")
     , layer_thickness
-    , (layer_nr == 0)? mesh.getSettingInPercentage("material_flow_layer_0") : mesh.getSettingInPercentage("material_flow")
-    , GCodePathConfig::SpeedDerivatives{mesh.getSettingInMillimetersPerSecond("speed_roofing"), mesh.getSettingInMillimetersPerSecond("acceleration_roofing"), mesh.getSettingInMillimetersPerSecond("jerk_roofing")}
+    , (layer_nr == 0)? mesh.settings.get<Ratio>("material_flow_layer_0") : mesh.settings.get<Ratio>("material_flow")
+    , GCodePathConfig::SpeedDerivatives{mesh.settings.get<Velocity>("speed_roofing"), mesh.settings.get<Velocity>("acceleration_roofing"), mesh.settings.get<Velocity>("jerk_roofing")}
 )
 , ironing_config(
     PrintFeatureType::Skin
-    , mesh.getSettingInMicrons("skin_line_width")
+    , mesh.settings.get<coord_t>("skin_line_width")
     , layer_thickness
-    , (layer_nr == 0)? mesh.getSettingInPercentage("material_flow_layer_0") : mesh.getSettingInPercentage("material_flow")
-    , GCodePathConfig::SpeedDerivatives{mesh.getSettingInMillimetersPerSecond("speed_ironing"), mesh.getSettingInMillimetersPerSecond("acceleration_ironing"), mesh.getSettingInMillimetersPerSecond("jerk_ironing")}
+    , (layer_nr == 0)? mesh.settings.get<Ratio>("material_flow_layer_0") : mesh.settings.get<Ratio>("material_flow")
+    , GCodePathConfig::SpeedDerivatives{mesh.settings.get<Velocity>("speed_ironing"), mesh.settings.get<Velocity>("acceleration_ironing"), mesh.settings.get<Velocity>("jerk_ironing")}
 )
 
 , perimeter_gap_config(createPerimeterGapConfig(mesh, layer_thickness, layer_nr))
 {
     infill_config.reserve(MAX_INFILL_COMBINE);
 
-    double flow = (layer_nr == 0) ? mesh.getSettingInPercentage("material_flow_layer_0") : mesh.getSettingInPercentage("material_flow");
+    Ratio flow = (layer_nr == 0) ? mesh.settings.get<Ratio>("material_flow_layer_0") : mesh.settings.get<Ratio>("material_flow");
 
     // Use the infill extruder's settings for the infill config if specified
-    const int infill_extruder_nr = mesh.getSettingAsIndex("infill_extruder_nr");
+    const int infill_extruder_nr = mesh.settings.get<int>("infill_extruder_nr");
```

Comment:
This is related to the bug described in CURA-5057. It was strangely only applied to infill. I'm transforming it only for infill now so that the remaining bug for CURA-5057 is fixed, but it should also be applied to all the rest of the line configs. I didn't do that though, because this is a workaround around an architectural defect which would be really bad to copy-paste and even worse since the rest of the configs are constructed via in-line expressions. I just fixed the one where there was already a bad workaround for.

Reply:
About that fix:
It took me quite some time to understand what was going wrong and I had an idea for making the fix more easy to read.
First recap the problem: a setting like 'flow' is reused by all different features types (infill, walls, etc), so it cannot be limited-to-extruder, but the flw on infill should still use the infill_extruder_nr.
My idea for a more clear fix:
`Mesh::settings.get<return_type>(setting_key_string, extruder_string)`
This function should return the setting if it is in the mesh storage, but otherwise use the limit-to-extruder provided as the second argument.

Hope you like my suggestion.

Reply:
On another note:

Please don't start fixing bugs on this branch as well!!
Please keep this branch limited in scope. If you start fixing things in this PR then you know the output of the program will not be exactly the same, but given that this PR touches almost all code that means that in order to test the engine after the refactor we have to manually check all the output whether it is correct, rather than testing whether the output is simply the same as before.

Please factor out any bugfixes into a separate PR.

Reply:
> a setting like 'flow' is reused by all different features types (infill, walls, etc), so it cannot be limited-to-extruder, but the flow on infill should still be obtained from the infill_extruder_nr.

Yes, that is the gist of the problem I ended up at as well. We've talked this over with the team and decided to fix it separately (but first think of the best solution).